### PR TITLE
Avoid TestContext write string allocations

### LINKED ISSUE #10095: [efficiency-improver] Avoid string allocation in TestContext.Write/WriteLine when no null chars present
## Goal and Rationale

Every call to `TestContext.Write(string)`, `TestContext.Write(string, params)`, `TestContext.WriteLine(string)`, and `TestContext.WriteLine(string, params)` unconditionally called `string.Replace("\\0", "\\\\0")`, which allocates a new `string` object even when the message contains no null characters.

Null characters in test output messages are rare edge cases. In the common case — the vast majority of calls — this Replace allocates a string identical to the input, which is immediately eligible for GC.

## Focus Area

**Code-Level Efficiency** — unnecessary object allocation / memory churn

## Approach

Extract a private `EscapeNullChars(string?)` helper that short-circuits on `Contains(char)` before calling `Replace`. The `Contains(char)` overload is O(n) but allocates nothing; `Replace` is O(n) and allocates a new string. When `\0` is absent, the helper returns the original reference with zero allocation.

```diff
- string? msg = message?.Replace("\0", "\\0");
+ string? msg = EscapeNullChars(message);

+ // Avoids a string allocation in the common case where there are no null characters.
+ private static string? EscapeNullChars(string? value)
+     => value is not null && value.Contains(nullChar) ? value.Replace("\0", "\\0") : value;
```

Applied to all four `Write`/`WriteLine` overloads.

## Energy Efficiency Evidence

**Proxy metric**: Memory allocation (string heap churn → GC pressure → CPU energy for collection)

**Baseline**: Every `TestContext.Write*` call allocates one extra `string` object (copy of the message). For a test suite that logs 1 000 messages per run, this is 1 000 unnecessary heap allocations.

**After**: Zero extra allocations when no `\0` characters present (the common case). The `Contains` check adds ~1 ns of CPU time vs. the allocation + GC cost saved.

**Methodology note**: Direct energy measurement is not possible in this environment. Proxy is heap allocation reduction, which reduces GC frequency and DRAM activity. This aligns with the GSF *Hardware Efficiency* principle — making better use of existing hardware by reducing unnecessary memory traffic.

**Limitation**: Impact scales with how frequently tests call `TestContext.Write*`. Suites that heavily use test context logging (e.g. data-driven tests writing per-row diagnostics) benefit most.

## Green Software Foundation Context

**Hardware Efficiency**: Fewer allocations → fewer GC pauses → lower CPU utilisation per test run → proportionally less energy per functional unit (test executed).

**SCI (Software Carbon Intensity)**: Reducing GC pressure lowers energy consumed per test run (E in the SCI equation), directly improving the per-functional-unit carbon score.

## Trade-offs

The `EscapeNullChars` helper adds a two-scan overhead (one `Contains`, one `Replace`) only when `\0` chars are present — a negligible cost for the already-rare path. Code readability is unchanged or slightly improved by extracting the helper.

## Reproducibility

To measure allocation reduction, profile a test run with `dotnet-trace` (Allocation event provider) before and after, comparing `System.String` object allocation counts in `TestContextImplementation.Write*` call stacks.

## Test Status

The repo-local SDK (`11.0.100-preview.7`) is not available in the agent environment, so a full build could not be run. The change is a purely mechanical textual substitution — introducing a private helper and calling it from the four existing sites — with no behavioural change in the null-char-present path.

> [!WARNING]
> <details>
> <summary>Firewall blocked 1 domain</summary>
>
> The following domain was blocked by the firewall during workflow execution:
>
> - `awmgmcpg`
>> To allow these domains, add them to the `network.allowed` list in your workflow frontmatter:
>
> ```yaml
> network:
>   allowed:
>     - defaults
>     - "awmgmcpg"
> ```
>
> See [Network Configuration](https://github.github.com/gh-aw/reference/network/) for more information.
>
> </details>

> 🤖 **Automated content by GitHub Copilot.** Generated by the [Efficiency Improver](https://github.com/microsoft/testfx/actions/runs/29782192334/agentic_workflow) workflow. · 336.2 AIC · ⌖ 10.2 AIC · ⊞ 11.7K · [◷]( · [◷](https://github.com/search?q=repo%3Amicrosoft%2Ftestfx+%22gh-aw-workflow-id%3A+efficiency-improver%22&type=pullrequests))
>
<details>
<summary><sub>Add this agentic workflow to your repo</sub></summary>

To install this agentic workflow, run

```
gh aw add githubnext/agentics/workflows/efficiency-improver.md@main
```
</details>

---

> [!NOTE]
> This was originally intended as a pull request, but GitHub Actions is not permitted to create or approve pull requests in this repository.
> The changes have been pushed to branch `efficiency/avoid-null-char-replace-in-testcontext-write-c0ab13c0bc15b562`.
>
> **[Click here to create the pull request](https://github.com/microsoft/testfx/compare/main...efficiency/avoid-null-char-replace-in-testcontext-write-c0ab13c0bc15b562?expand=1&title=%5Befficiency-improver%5D%20Avoid%20string%20allocation%20in%20TestContext.Write%2FWriteLine%20when%20no%20null%20chars%20present)**

To fix the permissions issue, go to **Settings** → **Actions** → **General** and enable **Allow GitHub Actions to create and approve pull requests**. See also: [gh-aw FAQ](https://github.github.com/gh-aw/reference/faq/#why-is-my-create-pull-request-workflow-failing-with-github-actions-is-not-permitted-to-create-or-approve-pull-requests)

<details><summary>Show patch preview (70 of 70 lines)</summary>

```diff
From 98eae2c504cc4d366ef7737abffce79a76de3d21 Mon Sep 17 00:00:00 2001
From: "github-actions[bot]" <github-actions[bot]@users.noreply.github.com>
Date: Mon, 20 Jul 2026 22:10:08 +0000
Subject: [PATCH] Avoid string allocation in TestContext.Write/WriteLine when
 no null chars present

Replace the unconditional string.Replace("\0", "\\0") calls in
TestContextImplementation.Write/WriteLine with a short-circuit helper
that only allocates when the input actually contains a null character.

In the common case (no \0 in the message), this eliminates one string
allocation per TestContext.Write/WriteLine call.

Co-authored-by: Copilot <223556219+Copilot@users.noreply.github.com>
---
 .../Services/TestContextImplementation.cs            | 12 ++++++++----
 1 file changed, 8 insertions(+), 4 deletions(-)

diff --git a/src/Adapter/MSTestAdapter.PlatformServices/Services/TestContextImplementation.cs b/src/Adapter/MSTestAdapter.PlatformServices/Services/TestContextImplementation.cs
index 12b9574..220c1aa 100644
--- a/src/Adapter/MSTestAdapter.PlatformServices/Services/TestContextImplementation.cs
+++ b/src/Adapter/MSTestAdapter.PlatformServices/Services/TestContextImplementation.cs
@@ -226,7 +226,7 @@ public override void AddResultFile(string fileName)
     /// <param name="message">The formatted string that contains the trace message.</param>
     public override void Write(string? message)
     {
-        string? msg = message?.Replace("\0", "\\0");
+        string? msg = EscapeNullChars(message);
         TestContextMessageBuilder.Append(msg);
         WriteLive(msg, appendLine: false);
     }
@@ -239,7 +239,7 @@ public override void Write(string? message)
     /// <param name="args">Arguments to add to the trace message.</param>
     public override void Write(string format, params object?[] args)
     {
-        string message = string.Format(CultureInfo.CurrentCulture, format.Replace("\0", "\\0"), args);
+        string message = string.Format(CultureInfo.CurrentCulture, Esc
... (truncated)
```

</details>

### PR DESCRIPTION
**Pull request by Evangelink**
## Summary

Avoid calling `string.Replace` in `TestContext.Write` and `WriteLine` when the input contains no null characters. The new helper returns the original string in the common case and preserves the existing escaping behavior when null characters are present.

Fixes #10095

## Testing

- Built `MSTestAdapter.PlatformServices.UnitTests` for `net8.0`
- Passed all 927 `MSTestAdapter.PlatformServices.UnitTests` tests

### CODE REVIEW

File: src/Adapter/MSTestAdapter.PlatformServices/Services/TestContextImplementation.cs [hunk 1]
```diff
@@ -264,11 +264,15 @@ public override void WriteLine(string? message)
...
     {
-        string message = string.Format(CultureInfo.CurrentCulture, format.Replace("\0", "\\0"), args);
+        string message = string.Format(CultureInfo.CurrentCulture, EscapeNullChars(format), args);
         TestContextMessageBuilder.AppendLine(message);
         WriteLive(message, appendLine: true);
     }
 
+    [return: NotNullIfNotNull(nameof(value))]
+    private static string? EscapeNullChars(string? value)
+        => value is not null && value.Contains('\0') ? value.Replace("\0", "\\0") : value;
```
[the remark this answers is not in the document]
Comment (Evangelink) [reply src/Adapter/MSTestAdapter.PlatformServices/Services/TestContextImplementation.cs]: Good catch — reverted the redundant pre-scan and helper in 2688f6619. `string.Replace` already returns the original instance when no match exists.

File: src/Adapter/MSTestAdapter.PlatformServices/Services/TestContextImplementation.cs [hunk 2]
```diff
@@ -264,11 +264,15 @@ public override void WriteLine(string? message)
     /// <param name="args">Arguments to add to the trace message.</param>
     public override void WriteLine(string format, params object?[] args)
     {
-        string message = string.Format(CultureInfo.CurrentCulture, format.Replace("\0", "\\0"), args);
+        string message = string.Format(CultureInfo.CurrentCulture, EscapeNullChars(format), args);
```
[the remark this answers is not in the document]
Comment (Evangelink) [reply src/Adapter/MSTestAdapter.PlatformServices/Services/TestContextImplementation.cs]: This alert is unrelated to the proposed change. Normalizing enumerable arguments would change the established `string.Format` behavior of `TestContext.WriteLine`; 2688f6619 restores the original implementation instead.

File: src/Adapter/MSTestAdapter.PlatformServices/Services/TestContextImplementation.cs [hunk 3]
```diff
@@ -239,7 +239,7 @@ public override void Write(string? message)
     /// <param name="args">Arguments to add to the trace message.</param>
     public override void Write(string format, params object?[] args)
     {
-        string message = string.Format(CultureInfo.CurrentCulture, format.Replace("\0", "\\0"), args);
+        string message = string.Format(CultureInfo.CurrentCulture, EscapeNullChars(format), args);
```
[the remark this answers is not in the document]
Comment (Evangelink) [reply src/Adapter/MSTestAdapter.PlatformServices/Services/TestContextImplementation.cs]: This alert is unrelated to the proposed change. Normalizing array arguments would change the established `string.Format` behavior of `TestContext.Write`; 2688f6619 restores the original implementation instead.

**Review by 0101 — APPROVED** [review]
